subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/ConfigurationContainerExtensions.kt

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+@file:Incubating
+
+
+package org.gradle.kotlin.dsl
+
+import org.gradle.api.Action
+import org.gradle.api.Incubating
+import org.gradle.api.NamedDomainObjectContainer
+import org.gradle.api.NamedDomainObjectProvider
+import org.gradle.api.PolymorphicDomainObjectContainer
+import org.gradle.api.artifacts.Configuration
+import org.gradle.api.artifacts.ConfigurationContainer
+import org.gradle.api.artifacts.ConsumableConfiguration
+import org.gradle.api.artifacts.DependenciesConfiguration
+import org.gradle.api.artifacts.Dependency
+import org.gradle.api.artifacts.ResolvableConfiguration
+import org.gradle.kotlin.dsl.support.delegates.NamedDomainObjectContainerDelegate
+import kotlin.reflect.KClass
+
+
+/**
+ * Allows the container to be configured via an augmented DSL.
+ *
+ * @param configuration The expression to configure this container with
+ * @return The container.
+ *
+ * @since 8.3
+ */
+@Suppress("nothing_to_inline")
+@Incubating
+inline operator fun ConfigurationContainer.invoke(
+    configuration: Action<ConfigurationContainerScope>
+): ConfigurationContainer = apply {
+    configuration.execute(ConfigurationContainerScope.of(this))
+}
+
+
+/**
+ * Receiver for [ConfigurationContainer] configuration blocks.
+ *
+ * @since 8.3
+ */
+@Incubating
+class ConfigurationContainerScope
```

Comment:
@gradle/bt-kotlin-dsl-maintainers PTAL, I thought we wanted to get away from these Scope-things, but I don't quite understand if we need it.

Reply:
@gradle/bt-kotlin-dsl-maintainers Thoughts here?

Reply:
We want to get rid of `*Scope`s eventually, yes.

IIUC this is added here to support the "String invoke" syntax only.
We have plans to support the "String invoke" syntax with Kotlin's context receivers instead.

Is there any tests for the String invoke syntax in this PR for this?
Is the String invoke syntax required for this PR? If not I would refrain to add this now.

Reply:
Consider the following buildscript
```
configurations {
    println(this)
}
```

Prior to adding this `ConfigurationContainerScope`, and the `invoke(Action)` method above, the buildscript output was:
```
org.gradle.kotlin.dsl.NamedDomainObjectContainerScope@3f172e7f
```

Without adding `ConfigurationContainerScope`, this code handles the configure action for the configuration container instead: https://github.com/gradle/gradle/blob/7ea983b74d67f415a3247657d4444fdbc8262e65/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt#L30-L41

The problem is -- is that `NamedDomainObjectContainerScope` does not implement the new factory methods that this PR is trying to add. 

While `ConfigurationContainerScope` does implement string invoke functionality, _so does the existing_ scope object. Taking that away would be a breaking change. 

Is there a way to allow users to call these factory methods in the closure without implementing a new `Scope`?